lib/hexstring.ml

```diff
@@ -1,17 +1,35 @@
 (* encoding bytearray -> hexstring *)
 
 let encode (bytearray : bytes) : string =
-  let rec encode_rec fold bytearray =
-    let length = Bytes.length bytearray in
-    if length = 0 then fold
+  let dummy_char = '_' in
+  let start_of_digit_0_in_ascii_table = 0x30 in
+  let start_of_lower_case_a_in_ascii_table = 0x61 in
+  let hex_digit_of_int (x : int) : char =
+    assert (x >= 0);
+    assert (x < 16);
+    char_of_int (
+      if x < 10 then
+        x + start_of_digit_0_in_ascii_table
+      else
+        x + start_of_lower_case_a_in_ascii_table
+    )
+  in
+  let rec aux bytearray len cur_pos buf =
+    if cur_pos < len then
+      let x = int_of_char @@ Bytes.get bytearray cur_pos in
+      let c1 = hex_digit_of_int (x / 0x10) in
+      let c2 = hex_digit_of_int (x mod 0x10) in
```

Comment:
I would have written that with `lsr 4` which seems more appropriate since we're manipulating bits and not math (the reader might also wonder here how the division behaves in case in cases where 0x10 doesn't divide x)

Reply:
Okay yep, `lsr 4` would have been nicer yep - just didn't think of shifting at all when I wrote it.

Reply:
I have replaced `x / 0x10` with `x lsr 4` and `x mod 0x10` with `x land 0x0F`